### PR TITLE
Introduce BuildInfo

### LINKED ISSUE #2106: Support for `BuildInfo`
**Is your feature request related to a problem? Please describe.**
Discussed in https://github.com/VirtusLab/scala-cli/discussions/2051

<div type='discussions-op-text'>

<sup>Originally posted by **ckipp01** April 23, 2023</sup>
I have a use case where I'm building a CLI app. A lot of times with apps like this you want to make sure you can display the version of the app via a command, and many times the value of that version is persisted utilizing a form of `BuildInfo` where based on git, or something else a file is generated and used for compilation. Is there any way to mimic something like this with `scala-cli`? All the solutions I'm coming up with request running `scala-cli` to generate a file first, and then running it again on your main project to utilize that created file. However, it'd be great if this could just be done in a single step somehow. Any ideas?</div>

**Describe the solution you'd like**
We could have a directive/command-line option exposing `BuildInfo` data in the project.
```scala
//> using build.info
```

**Describe alternatives you've considered**
This could potentially also be delivered through code generators. (see #610)

**Additional context**
See the discussions in #2051 and #610.

### PR DESCRIPTION
**Pull request by MaciejG604**
Addresses #2106

As well as moves the wrapper generation to the step when Sources are created from ScopedSources

### CODE REVIEW

**Comment by lwronski** [comment]
Based on discussion in https://github.com/VirtusLab/scala-cli/discussions/2051, I think that  the `BuildInfo` should include the project's version.

File: modules/build/src/main/scala/scala/build/ScopedSources.scala [hunk 1]
```diff
@@ -17,12 +36,20 @@ final case class ScopedSources(
     }
 
   def sources(scope: Scope, baseOptions: BuildOptions): Sources =
+    val combinedBuildOptions = buildOptionsFor(scope)
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/ScopedSources.scala]: Could you include comments indicating the origin of the `baseOptions` and `buildOptions`?

File: modules/build/src/main/scala/scala/build/ScopedSources.scala [hunk 2]
```diff
@@ -36,20 +37,55 @@ final case class ScopedSources(
...
       .flatMap(_.valueFor(scope).toSeq)
       .map(_.wrap(codeWrapper))
 
+    val needsBuildInfo = combinedOptions.sourceGeneratorOptions.useBuildInfo.getOrElse(false)
+
+    val maybeBuildInfoSource = if (needsBuildInfo && scope == Scope.Main)
+      Seq(Sources.InMemory(
+        Left("build-info"),
+        os.rel / "BuildInfo.scala",
+        buildInfo(baseOptions).generateContents(),
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/ScopedSources.scala]: Why aren't you using `combinedOptions` in this case?

Comment (MaciejG604) [reply modules/build/src/main/scala/scala/build/ScopedSources.scala]: Good catch! TY

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala [hunk 3]
```diff
@@ -137,6 +137,19 @@ class BspServer(
...
+      .filter(_.reportingPath.isLeft)
+
+    for {
+      item <- res.getItems.asScala
+      if validTarget(item.getTarget)
+      sourceItem <- item.getSources.asScala
+      if sourcesWithReportingPathString.exists(
+        _.generated.toNIO.toUri.toASCIIString == sourceItem.getUri
+      )
+    } sourceItem.setGenerated(true)
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspServer.scala:152]: I was confused regarding to the type of `reportingPath`. I think it shouldn't be an `Either` type but rather, it could be better represented by a case class with Scala doc explaining the type of path.

But I think It can be added in the new following PR.

**Review by lwronski — COMMENTED** [review]
Adding integration test for this feature would be beneficial.

File: modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala [hunk 4]
```diff
@@ -1488,4 +1488,82 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
...
+           |//> using resourceDir ./resources
+           |//> using jar TEST1.jar TEST2.jar
+           |
+           |//> using buildInfo
+           |
+           |import scala.cli.build.BuildInfo
+           |
+           |object Main extends App {
+           |  val buildInfoChecks = Seq(
+           |   BuildInfo.scalaVersion == "$actualScalaVersion",
```
Comment (lwronski) [inline modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala]: It can be used `assert` in the code to verify the condition.

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/ComputeVersion.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  HasBuildOptions
+}
+import scala.build.errors.BuildException
+import scala.build.options.{BuildOptions, ComputeVersion => cv, SourceGeneratorOptions}
+import scala.cli.commands.SpecificationLevel
+
+@DirectiveGroupName("Compute Version")
+@DirectiveExamples("//> using computeVersion git:tag")
+@DirectiveExamples("//> using computeVersion command:'echo 1.0.0'")
+@DirectiveUsage("//> using computeVersion git:tag", "`//> using platform `git:tag")
```
Comment (tgodzik) [inline modules/directives/src/main/scala/scala/build/preprocessing/directives/ComputeVersion.scala]: What is ` using platform `git:tag`  used for?  Could we actually separate computeVersion into a separate PR?

File: modules/build/src/test/scala/scala/build/tests/SourceGeneratorTests.scala [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+             |  val platform = "JVM"
+             |  val jvmVersion = Some("11")
+             |  val scalaJsVersion = None
+             |  val jsEsVersion = None
+             |  val scalaNativeVersion = None
+             |  val mainClass = Some("Main")
+             |  val projectVersion = Some("TestVersion")
+             |
+             |
+             |  object MainScopeInfo {
```
Comment (tgodzik) [inline modules/build/src/test/scala/scala/build/tests/SourceGeneratorTests.scala]:
```suggestion
             |  object Main {
```
wouldn't `Main` and `Test` be enough? `BuildInfo.Main` already seems clear. Or at least drop the `Info` suffix

Comment (lwronski) [reply modules/build/src/test/scala/scala/build/tests/SourceGeneratorTests.scala]: I think that `BuildInfo.Main` should be sufficient.

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/ComputeVersion.scala (hunk 5 above)
Comment (MaciejG604) [reply modules/directives/src/main/scala/scala/build/preprocessing/directives/ComputeVersion.scala]: `platform` is a mistake here. I could extract the commits with computeVersion to a different PR, I'll do that

**Comment by MaciejG604** [comment]
ComputeVersion is now moved to a different branch and will be a follow-up PR to this one
https://github.com/MaciejG604/scala-cli/pull/4

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/BuildInfo.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+@DirectiveExamples("//> using buildInfo")
+@DirectiveUsage("//> using buildInfo", "`//> using buildInfo")
+@DirectiveDescription("Generate BuildInfo for project")
+@DirectiveLevel(SpecificationLevel.RESTRICTED)
+final case class BuildInfo(
+  buildInfo: Boolean = false
+) extends HasBuildOptions {
+  def buildOptions: Either[BuildException, BuildOptions] =
+    val options = BuildOptions(sourceGeneratorOptions =
+      SourceGeneratorOptions(useBuildInfo = Some(true))
```
Comment (lwronski) [inline modules/directives/src/main/scala/scala/build/preprocessing/directives/BuildInfo.scala]:
```suggestion
      SourceGeneratorOptions(useBuildInfo = Some(buildInfo))
```

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala (hunk 3 above)
Comment (lwronski) [reply modules/build/src/main/scala/scala/build/bsp/BspServer.scala:152]: Can you create an issue to ensure we don't forget about this?

**Review by lwronski — APPROVED** [review]
LGTM

File: modules/build/src/main/scala/scala/build/Build.scala [hunk 8]
```diff
@@ -267,9 +267,7 @@ object Build {
 
       val baseOptions = overrideOptions.orElse(sharedOptions)
 
-      val wrappedScriptsSources = crossSources.withWrappedScripts(baseOptions)
```
Comment (tgodzik) [inline modules/build/src/main/scala/scala/build/Build.scala:270]: Quick question -> why do we not need to add scripte sources here anymore?

Comment (tgodzik) [reply modules/build/src/main/scala/scala/build/Build.scala:270]: Ach nvm, I see

File: modules/build/src/main/scala/scala/build/bsp/BspServer.scala (hunk 3 above)
Comment (MaciejG604) [reply modules/build/src/main/scala/scala/build/bsp/BspServer.scala:152]: #2308

**Review by tgodzik — APPROVED** [review]
LGTM!
